### PR TITLE
Contributing valkeyJSON module

### PR DESCRIPTION
**Pull request by roshkhatri**
This PR adds the implementation of valkeyJSON module based on the specifications detailed in the accepted RFC document: [ValkeyJSON Specification](https://github.com/valkey-io/valkey-rfc/blob/main/ValkeyJSON.md).

The implementation does not include JSON.MSET implementation but I am working on it as we review this. I am hoping we could add the implementation post merging this PR.

@valkey-io/core-team

### CODE REVIEW

File: src/json/json_api.h
Comment (madolson) [inline src/json/json_api.h]: What's a search module?

**Comment by hwware** [comment]
Sorry，what's the blocker for  JSON.MSET implementation? ^_^?

**Comment by roshkhatri** [comment]
> Sorry，what's the blocker for JSON.MSET implementation? ^_^?

There is no blocker. its WIP. That can be added in a separate PR.

File: CMakeLists.txt
Comment (zuiderkwast) [inline CMakeLists.txt:75]: I would like to see some way to not download dependencies during the build, at least optionally.

Offline builds are common to guarantee that a project doesn't have any online dependencies that have not been verified in advance. Supply chain attacks is the main concern. License/legal problems is another.

I suggest we add some environmant variable to provide the dependencies using a local paths and CMake can use these instead of downloading them if they are provided. It can be added in the future though.

**Review by zuiderkwast — APPROVED** [review]
There is no way that I can review all this. 27K in one commit. :D

There will need to be maintainer team for each module that can review each other's PRs. So, at least two maintainers would be useful. Is there anyone besides you @roshkhatri?

I'd say let's merge it! Later, we can talk about beta releases and stable releases. The TSC can oversee your release plans and the work on a high level, but I don't see that TSC members have enough time to be deeply involved in these modules.

File: CMakeLists.txt
Comment (zuiderkwast) [reply CMakeLists.txt:75]: > I would like to see some way to not download dependencies during the build, at least optionally.
> 
> Offline builds are common to guarantee that a project doesn't have any online dependencies that have not been verified in advance. Supply chain attacks is the main concern. (License issues is another concern.)
> 
> I suggest we add some environmant variable to provide the dependencies using a local paths and CMake can use these instead of downloading them if they are provided. It can be added in the future though.

File: src/json/json_api.h
Comment (joehu21) [reply src/json/json_api.h]: Let's just say "JSON C API".

**Comment by roshkhatri** [comment]
>There will need to be maintainer team for each module that can review each other's PRs. So, at least two maintainers would be useful. Is there anyone besides you @roshkhatri?

Yes, @joehu21  will be the maintainer besides me.

>I'd say let's merge it! Later, we can talk about beta releases and stable releases. The TSC can oversee your release plans and the work on a high level, but I don't see that TSC members have enough time to be deeply involved in these modules.

Yes, once this is merged we can open issues and discuss/work on them.

**Comment by hwware** [comment]
Mostly it looks like fine, agree to merge it. Great work.

**Review by hwware — APPROVED** [review]

**Review by joehu21 — APPROVED** [review]
LGTM

File: src/json/json.cc
Comment (madolson) [inline src/json/json.cc:2540]: The RFC has 2 configs, why does this have like 50?

File: src/json/stats.h
Comment (madolson) [inline src/json/stats.h:4]: The original rfc has 2 metrics, why does this one have so much more?

**Review by madolson — COMMENTED** [review]
Just evaluated the RFC against the implementation. There seems to be some discrepancies, so would like to start conversations about why.

File: src/json/json.cc
Comment (madolson) [inline src/json/json.cc:2652]: I suppose I don't get what this is for. What does this do for compatibility?

Comment (madolson) [inline src/json/json.cc:2652]: I suppose I don't get what this is for. What does this do for compatibility?

File: src/json/json.cc
Comment (madolson) [inline src/json/json.cc]: Overall, this code is not very well documented. This comment just trails off for example.

File: src/json/json.cc
Comment (madolson) [inline src/json/json.cc:2830]: Apparently we also load 3 custom module types here. I understand this to be for compatibility with stack, but do we want to have this in this module?

File: src/json/json.cc
Comment (madolson) [inline src/json/json.cc:2906]: This code doesn't look linted, maybe create a followup issue to make sure we actually running the linter automatically. I guess generally we're missing workflow actions.

**Review by madolson — COMMENTED** [review]
I am mostly trusting the code, can we setup github workflows to make sure this code all builds and what not?

File: src/json/json.cc
Comment (madolson) [inline src/json/json.cc]: 👻 Mystery file in API.md.

**Comment by madolson** [comment]
Ideally let's make issues for the comments here. The only thing I think we should fix before merging is getting the CI working.

File: src/json/json.cc
Comment (zuiderkwast) [reply src/json/json.cc:2540]: 5️⃣0️⃣⁉️

It's easy to add configs but it's a breaking change to remove one. Without looking at what they are for, can we replace them with constants or something before we release a version where we commit to maintaining them?

File: src/json/json.cc
Comment (zuiderkwast) [reply src/json/json.cc:2830]: Interesting. At least, this should be added to the rfc.

**Comment by roshkhatri** [comment]
I have added the CI workflow to run against valkey `unstable` and `8.0.0` for now.

Screenshot 2024-11-26 at 12 57 52 PM

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2830]: Yes, that's for RDB compatibility with RedisJSON. If we don't care about them any more, we can remove the code.

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2652]: The function is invoked when loading a RDB produced by RedisJSON. Their RDB could include a newer version of a module type. This function decides whether or not to reject the RDB.

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2540]: We considered dropping these configs. However, it requires a decent amount of code cleanup. Agreed we should remove them. We can create an issue to replace these configs with constants.

File: src/json/json.cc
Comment (madolson) [reply src/json/json.cc:2652]: Won't it just get rejected if it's from a newer version?

File: src/json/stats.h
Comment (joehu21) [reply src/json/stats.h:4]: These module infos have been removed (not emitted). However, the documentation is forgotten to cleaned up.

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc]: Yeah, the comment should be "Check an integer value, fail if the value doesn't match the expected one."

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc]: The line should be removed.

File: src/json/json.cc
Comment (zuiderkwast) [reply src/json/json.cc:2830]: I think we do care about compatibility with RedisJSON. We want to make it easy to migrate to Valkey.

I just wish it is mentioned in the RFC, which already has a lot of details.

Comment (madolson) [reply src/json/json.cc:2830]: These aren't for json compatibility though, they list gears and search types.

Comment (zuiderkwast) [reply src/json/json.cc:2830]: Which are partially used by the JSON module or vice versa? Or the JSON module provides a minimal subset of them or something? I don't know this. :laughing:

I guess the only reason to add them is to support RDB streams containing module types from these modules. Otherwise, delete it.

File: .github/workflows/ci.yml
Comment (PingXie) [inline .github/workflows/ci.yml:13]: consider using the commit hash for the action instead.

File: CMakeLists.txt
Comment (PingXie) [reply CMakeLists.txt:75]: looks like this is all about reusing the tests at https://github.com/valkey-io/valkey-bloom/tree/unstable/tests? or there is more?

File: CMakeLists.txt
Comment (PingXie) [inline CMakeLists.txt]: print out the architecture?

File: .github/workflows/ci.yml
Comment (PingXie) [inline .github/workflows/ci.yml:11]: should we go with the tip of the release branch, which is 8.0? Also I think having older engine versions (like 7.2) at some point would help catch regressions. worst case though we could end up with 6 versions based on our support promise.

File: CMakeLists.txt
Comment (PingXie) [inline CMakeLists.txt]: not going with -O3?

File: CMakeLists.txt
Comment (PingXie) [inline CMakeLists.txt]: nit - print out the actual arch too?

File: CMakeLists.txt
Comment (PingXie) [inline CMakeLists.txt]: btw, this comment can be skipped IMO
```suggestion
# Add the tst subdirectory for building
```

File: README.md
Comment (PingXie) [inline README.md]: this motivation belongs more to the rfc. Developers who got here would probably care more about the how-tos IMO.

File: build.sh
Comment (PingXie) [inline build.sh:17]: this version check seems too restrictive. this tool is developer facing so might just open it up to all versions/commits

File: src/json/json.cc
Comment (PingXie) [reply src/json/json.cc:2830]: yeah these aux data types seem to just play a supporting role so far. can we register a generic stub instead and just log a message that "xxx is skipped" and move on? there could be more aux types added in the future in rejson so it will be good to have that visibility if community members report any issue.

```c
install_stub(ctx, "*", GENERIC_ENCVER, generic_aux_load);
```

File: src/json/json.cc
Comment (PingXie) [inline src/json/json.cc:2706]: these checks add fragility only to the module compat. Since we are not planning to consume any of them, the best strategy imo would be to log a "xxx is skipped" notice level message and return _OK. same goes to other stubs.

**Review by PingXie — APPROVED** [review]
This does look overwhelming and I only reviewed haphazardly. I agree we should just merge this PR and polish it incrementally, including adding more test coverage.

Thanks @roshkhatri and @joehu21!

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2652]: Default value of enforce_rdb_version_check is false. So, the check will pass. In our internal RDB compatibility tests, which isn't part of the PR, we set enforce_rdb_version_check to true so that we will be alerted when ReJSON introduced some new aux types. In retrospect, I think the aux type version checking is an overkill. We should just read and ignore these aux types.

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2830]: > These aren't for json compatibility though, they list gears and search types.

When testing compatibility with ReJSON, we use docker image "docker run -p 6389:6379 redis/redis-stack:latest", which loads a bunch of modules including search, graph, etc. Although our test only creates json keys, the generated RDB contains a bunch of aux types like scdtype00, gear*, etc. I suspect these aux types are from RedisSearch module, but not 100% sure because I never looked at their source code. (NOTE: It also indicates that their modules do not use aux_save2, which would not save aux types that have no corresponding data.)

To be able to load the RDB, we register these aux types, read the data and ignored them. So, these are indeed for JSON compatibility test.

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2706]: > yeah these aux data types seem to just play a supporting role so far. can we register a generic stub instead and just log a message that "xxx is skipped" and move on? there could be more aux types added in the future in rejson so it will be good to have that visibility if community members report any issue.

The challenge is that each aux type may have different number of entries in RDB. For example, scdtype only has one integer, while gearsdt has 7 entries in RDB (some are int and some are string).

File: CMakeLists.txt
Comment (joehu21) [reply CMakeLists.txt]: We want the max optimization that still gives us debug symbols with line numbers in the crash stack. I believe such opt level is -O2.

File: CMakeLists.txt
Comment (joehu21) [reply CMakeLists.txt]: Yes, printing out `${ARCHITECTURE}` will be helpful.

Comment (joehu21) [reply CMakeLists.txt]: > looks like this is all about reusing the tests at https://github.com/valkey-io/valkey-bloom/tree/unstable/tests? or there is more?

Yes, it's for reusing the pytest infrastructure in valkey-bloom.

File: CMakeLists.txt
Comment (joehu21) [reply CMakeLists.txt:75]: Yes, it's all about reusing the pytest infrastructure in valkey-bloom.

File: src/json/json.cc
Comment (PingXie) [reply src/json/json.cc:2830]: You are right @joehu21. These aux data types are indeed needed. The main purpose of these handlers is actually to skip these aux types.

File: CMakeLists.txt
Comment (PingXie) [reply CMakeLists.txt]: hmm this dependency is a bit odd - does it make sense to pull the pytest infra into the Valkey repo instead? We are discussing migrating to python based integration test infra for Valkey too.

File: CMakeLists.txt
Comment (PingXie) [reply CMakeLists.txt]: maybe at least make is configurable on the command line so it could be overwritten if so desired?

File: src/json/json.cc
Comment (PingXie) [reply src/json/json.cc:2706]: Got it - the parsing part makes sense to me now. But maybe we could skip the check of the actual value?

File: src/json/json.cc
Comment (zuiderkwast) [reply src/json/json.cc:2830]: If we ever want another module that provides compatibilty with redis search or gears modules, then these names are already registered by ValkeyJSON module so the other modules can't be loaded? That's a potential future problem.

File: CMakeLists.txt
Comment (roshkhatri) [reply CMakeLists.txt]: Yes, I would suggest we should pull it out in the valkey repo instead. However, we can change the line later when we have it separately.

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2706]: Totally agreed, no need to check the actual value.

File: CMakeLists.txt
Comment (joehu21) [reply CMakeLists.txt]: agreed. it's desirable to be able to override it through command line argument. @roshkhatri can you create an issue after the PR is merged?

File: src/json/json.cc
Comment (joehu21) [reply src/json/json.cc:2830]: That's a good point.

**Comment by madolson** [comment]
Before we merge, please just make sure everything has issues so we can track them.

**Review by madolson — APPROVED** [review]

**Review by enjoy-binbin — APPROVED** [review]
did not review, but it is indeed a good work, thanks.

File: CMakeLists.txt
Comment (madolson) [reply CMakeLists.txt]: There are still debug symbols with a crash at -O3, you might be thinking about frame pointers which allow unwinding the stack. You can keep those with `-fno-omit-frame-pointer`.

> @roshkhatri can you create an issue after the PR is merged?

Please create the issues before the PR is merged. It's better practice to define the follow up work before merging in the current code.

File: CMakeLists.txt
Comment (madolson) [reply CMakeLists.txt]: I was going to suggest we write an RFC or something similar for the python testing framework as well before we get too far along building it organically.

File: src/json/json.cc
Comment (madolson) [reply src/json/json.cc:2830]: Yeah, I think we need a better overall strategy here instead of install random types and ignoring the data from them.

File: src/json/json.cc
Comment (roshkhatri) [reply src/json/json.cc:2540]: https://github.com/valkey-io/valkeyJSON/issues/2

File: src/json/json.cc
Comment (roshkhatri) [reply src/json/json.cc:2906]: https://github.com/valkey-io/valkeyJSON/issues/3

Comment (roshkhatri) [reply src/json/json.cc:2906]: Added the CI to run all the tests for now.

File: .github/workflows/ci.yml
Comment (roshkhatri) [reply .github/workflows/ci.yml:13]: https://github.com/valkey-io/valkeyJSON/issues/4

File: .github/workflows/ci.yml
Comment (roshkhatri) [reply .github/workflows/ci.yml:11]: https://github.com/valkey-io/valkeyJSON/issues/5

File: CMakeLists.txt
Comment (roshkhatri) [reply CMakeLists.txt]: https://github.com/valkey-io/valkeyJSON/issues/6

File: build.sh
Comment (roshkhatri) [reply build.sh:17]: https://github.com/valkey-io/valkeyJSON/issues/7

File: src/json/json.cc
Comment (roshkhatri) [reply src/json/json.cc:2706]: https://github.com/valkey-io/valkeyJSON/issues/8

File: src/json/json.cc
Comment (roshkhatri) [reply src/json/json.cc:2830]: https://github.com/valkey-io/valkeyJSON/issues/9

**Comment by roshkhatri** [comment]
I have created relevant issues for the comments and resolved those conversations to track them better. I have also addressed some feedback for minor changes.

Let me know if any more issues are required to be opens before we merge this PR. Happy Thanksgiving Everyone!!

**Review by soloestoy — APPROVED** [review]

File: CMakeLists.txt
Comment (zuiderkwast) [reply CMakeLists.txt]: Why not move the testing framework to it's own repo to start with? It only needs to be pulled in for running the tests, not when just building the module.

I'd rather see people spend time on a README for the test framework than on an RFC. Or we can do both in parallel...?

**Comment by hwware** [comment]
> I have created relevant issues for the comments and resolved those conversations to track them better. I have also addressed some feedback for minor changes.
> 
> Let me know if any more issues are required to be opens before we merge this PR. Happy Thanksgiving Everyone!!

Happy Black Friday !

File: CMakeLists.txt
Comment (madolson) [reply CMakeLists.txt]: > I'd rather see people spend time on a README for the test framework than on an RFC. Or we can do both in parallel...?

I just want us to have some long term plan for testing and not just build it up organically. The reason is that Amazon internally has a similar testing framework that is extremely complex because nobody really maintains it and just bolt on random features. There are many cases where it's possible to do the same way 2 or 3 different ways. I think the same is true of our TCL framework, just to a smaller scale. 

So whether it be an RFC or just some direction in a new repo, I don't care which.

File: CMakeLists.txt
Comment (madolson) [reply CMakeLists.txt:75]: https://github.com/valkey-io/valkeyJSON/issues/10

File: src/json/stats.h
Comment (madolson) [reply src/json/stats.h:4]: https://github.com/valkey-io/valkeyJSON/issues/11

File: src/json/json.cc
Comment (madolson) [reply src/json/json.cc:2652]: https://github.com/valkey-io/valkeyJSON/issues/12

**Comment by madolson** [comment]
We have 6/6 TSC actually approved, so going ahead and merging it.

**Pull request merged**
